Port to ros2

vatanaksoytezer opened the pull request.

vatanaksoytezer commented:
This is currently blocked by: https://github.com/PickNikRobotics/rviz_visual_tools/pull/185 . With that merged this should work without any problems.

vatanaksoytezer commented:
Sorry, for forgetting this I will also need a ROS2 branch on https://github.com/PickNikRobotics/graph_msgs and send the port right there

henningkayser commented on src/imarker_robot_state.cpp:
```diff
@@ -108,39 +114,40 @@ bool isIKStateValid(const planning_scene::PlanningScene* planning_scene, bool ve
   visual_tools_ = std::allocate_shared<moveit_visual_tools::MoveItVisualTools>(
-      Eigen::aligned_allocator<moveit_visual_tools::MoveItVisualTools>(), psm_->getRobotModel()->getModelFrame(),
-      nh_.getNamespace() + "/" + imarker_name, psm_);
+      Eigen::aligned_allocator<moveit_visual_tools::MoveItVisualTools>(), node, psm_->getRobotModel()->getModelFrame(), "/" + imarker_name, psm_);
```
Topics starting with a `/` are global. So if we want to have them in the node's namespace (I think we do), then the slash should be dropped.

henningkayser commented on src/imarker_robot_state.cpp:
```diff
@@ -108,39 +114,40 @@ bool isIKStateValid(const planning_scene::PlanningScene* planning_scene, bool ve
+  // ! Add node here
+  // imarker_server_ = std::make_shared<interactive_markers::InteractiveMarkerServer>(imarker_topic, "", false);
+  // TODO: Check topic ns or imarker_topic
+  imarker_server_ = std::make_shared<interactive_markers::InteractiveMarkerServer>("", node);
```
Seems like the `imarker_topic` is not applied?

henningkayser replied:
So the topic namespace should still be used, the [constructor](https://github.com/ros-visualization/interactive_markers/blob/ros2/src/interactive_marker_server.cpp#L48) accepts that as well. I think it's still pretty important that namespaces are used everywhere, also in line 130 above.

henningkayser commented on src/imarker_robot_state.cpp:
```diff
@@ -41,14 +41,19 @@
 #include <vector>
 
+// Boost
+#include <boost/filesystem.hpp>
```
What is this used for?

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -35,10 +35,10 @@
 // Desc:   Simple tools for showing parts of a robot in Rviz, such as the gripper or arm
 
-#include <moveit_visual_tools/moveit_visual_tools.h>
+#include <moveit_visual_tools/moveit_visual_tools.hpp>
```
hmm.. we haven't decided to switch to `.hpp` for MoveIt, yet. At some point we might want to, but for now we should keep it consistent and stick to `.h`.

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -96,48 +103,51 @@ bool MoveItVisualTools::loadPlanningSceneMonitor()
   // Create tf transform buffer and listener
-  std::shared_ptr<tf2_ros::Buffer> tf_buffer = std::make_shared<tf2_ros::Buffer>();
-  std::shared_ptr<tf2_ros::TransformListener> tf_listener = std::make_shared<tf2_ros::TransformListener>(*tf_buffer);
+  // std::shared_ptr<tf2_ros::Buffer> tf_buffer = std::make_shared<tf2_ros::Buffer>();
```
Please remove

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -96,48 +103,51 @@ bool MoveItVisualTools::loadPlanningSceneMonitor()
+  // std::shared_ptr<tf2_ros::Buffer> tf_buffer = std::make_shared<tf2_ros::Buffer>();
+  // std::shared_ptr<tf2_ros::TransformListener> tf_listener = std::make_shared<tf2_ros::TransformListener>(*tf_buffer);
+
+  std::shared_ptr<tf2_ros::Buffer> tf_buffer =
```
Be aware of https://github.com/ros-planning/moveit2/pull/310, maybe add a TODO/Issue to remove the listener once that is merged

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -96,48 +103,51 @@ bool MoveItVisualTools::loadPlanningSceneMonitor()
+      std::make_shared<tf2_ros::Buffer>(node_->get_clock(), tf2::durationFromSec(10.0));
+  std::shared_ptr<tf2_ros::TransformListener> tfl = std::make_shared<tf2_ros::TransformListener>(*tf_buffer);
 
   // Regular version b/c the other one causes problems with recognizing end effectors
```
... I wonder what this was for

vatanaksoytezer replied:
I don't have much idea either on that one.

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -96,48 +103,51 @@ bool MoveItVisualTools::loadPlanningSceneMonitor()
-  ros::spinOnce();
-  ros::Duration(0.1).sleep();
-  ros::spinOnce();
+  rclcpp::spin_some(node_);
```
... and this... I don't think we should spin the node explicitly at all. This looks like a hack to work around some initialization issues. I would remove this and see if there are any problems. Most certainly, we can fix any syncing issues by waiting for the corresponding events using the PSM directly.

vatanaksoytezer replied:
You are right, I removed unnecessary spin_somes.

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -343,12 +347,13 @@ void MoveItVisualTools::loadTrajectoryPub(const std::string& display_planned_pat
   // Wait for topic to be ready
   if (blocking)
-    waitForSubscriber(pub_display_path_);
+    waitForSubscriber(pub_display_path_, 5.0);
```
This waits for 5 seconds, right? Please use inline comments when adding magic numbers, maybe like this:
```suggestion
    waitForSubscriber(pub_display_path_, 5.0 /* seconds */);
```
But should we even change the original behavior without proper error handling?

vatanaksoytezer replied:
I'll add the comments. I didn't get what you mean by the original behavior?

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -391,11 +397,11 @@ bool MoveItVisualTools::publishEEMarkers(const geometry_msgs::Pose& pose, const
 
     // Header
-    ee_markers_map_[ee_jmg].markers[i].header.stamp = ros::Time::now();
+    ee_markers_map_[ee_jmg].markers[i].header.stamp = rclcpp::Clock(RCL_ROS_TIME).now();
```
```suggestion
    ee_markers_map_[ee_jmg].markers[i].header.stamp = node_->now();
```

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -416,57 +422,57 @@ bool MoveItVisualTools::publishEEMarkers(const geometry_msgs::Pose& pose, const
     publishEEMarkers(possible_grasps[i].grasp_pose.pose, ee_jmg);
 
-    ros::Duration(animate_speed).sleep();
+    rclcpp::sleep_for(std::chrono::milliseconds(int(animate_speed * 1000)));
```
Maybe a better readable alternative:
```suggestion
    rclcpp::sleep_for(rclcpp::Duration::from_seconds(animate_speed));
```
Also, please extend the documentation that `animate_speed` is expecting seconds.
If we want to be really fancy, `animate_speed` should be a chrono type, but that is strictly out of scope of this PR. I'm thinking that this would be a good cleanup for the future, though.

vatanaksoytezer replied:
This also does not compile

vatanaksoytezer replied:
Add a small statement in doxy docs on animate speed should be given in seconds.

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -569,15 +575,15 @@ bool MoveItVisualTools::publishIKSolutions(const std::vector<trajectory_msgs::Jo
   {
     trajectory_pt_timed = ik_solutions[i];
-    trajectory_pt_timed.time_from_start = ros::Duration(running_time);
+    trajectory_pt_timed.time_from_start = rclcpp::Duration(running_time);
```
Doesn't really matter since `running_time` is `0.0`, but also here it's unclear what `running_time` really is and if the constructor is called this way intentionally. I'd prefer `rclcpp::Duration::from_seconds(running_time)`.

vatanaksoytezer replied:
`running_time` is incremented by `display_time`.

vatanaksoytezer replied:
I've added  `rclcpp::Duration::from_seconds(running_time)`

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -597,82 +603,82 @@ bool MoveItVisualTools::removeAllCollisionObjects()
-  moveit_msgs::CollisionObject co;
-  co.header.stamp = ros::Time::now();
+  moveit_msgs::msg::CollisionObject co;
+  co.header.stamp = rclcpp::Clock(RCL_ROS_TIME).now();
```
```suggestion
  co.header.stamp = node_->now();
```

henningkayser replied:
same for below

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -1261,45 +1267,45 @@ bool MoveItVisualTools::publishTrajectoryPath(const moveit_msgs::RobotTrajectory
     {
       counter += CHECK_TIME_INTERVAL;
-      ros::Duration(CHECK_TIME_INTERVAL).sleep();
+      rclcpp::sleep_for(std::chrono::milliseconds(int(CHECK_TIME_INTERVAL * 1000)));
```
```suggestion
      rclcpp::sleep_for(rclcpp::Duration::from_seconds(CHECK_TIME_INTERVAL));
```

vatanaksoytezer replied:
This does not compile, sleep_for cannot be initialized with `rclcpp::Duration::from_seconds`

henningkayser replied:
`std::chrono::seconds`?

vatanaksoytezer replied:
std::chrono::seconds takes integers only and I have 0.25 seconds to sleep.

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -1605,10 +1611,10 @@ planning_scene_monitor::PlanningSceneMonitorPtr MoveItVisualTools::getPlanningSc
-    ros::spinOnce();
-    ros::Duration(1).sleep();  // TODO: is this necessary?
+    rclcpp::spin_some(node_);
+    rclcpp::sleep_for(std::chrono::milliseconds(1000));  // TODO: is this necessary?
```
```suggestion
    rclcpp::sleep_for(1s);  // TODO: is this necessary?
```
using chrono literals.
Also, it should really not be necessary, `loadPlanningSceneMonitor()` should provide a fully initialized PSM

vatanaksoytezer replied:
This is removed as spin_some was not necessary.

henningkayser reviewed (commented):
Looks pretty clean so far :+1: Were you able to get all of the demo work as expected? I'll probably test this later today

vatanaksoytezer commented:
> Looks pretty clean so far  Were you able to get all of the demo work as expected? I'll probably test this later today.

@henningkayser Yes! Demo should work as expected. Thanks a lot for the detailed review, that is really helpful for me. I will send the fixes in Sunday when I get back home.

vatanaksoytezer commented:
One note for demo you would need the current foxy-devel branch of rviz_visual_tools and https://github.com/PickNikRobotics/graph_msgs/pull/5. And it should work. Let me know if something unexpected happens.

vatanaksoytezer commented on src/moveit_visual_tools.cpp:
```diff
@@ -1261,45 +1267,45 @@ bool MoveItVisualTools::publishTrajectoryPath(const moveit_msgs::RobotTrajectory
-  pub_display_path_.publish(display_trajectory_msg);
-  ros::spinOnce();
+  pub_display_path_->publish(display_trajectory_msg);
+  rclcpp::spin_some(node_);
```
I think this is meant to be specifically written like this. See developer notes in README.

vatanaksoytezer commented on CMakeLists.txt:
```diff
@@ -1,96 +1,116 @@
+install(DIRECTORY launch resources DESTINATION share/${PROJECT_NAME})
 
 # Install executables
 install(TARGETS ${PROJECT_NAME}_demo
```
This seemed more clear to me, and it seems we used a similar approach on most moveit2 packages. I can still do it this way if you think that would be more compact

vatanaksoytezer commented on CMakeLists.txt:
```diff
@@ -1,96 +1,116 @@
+  std_msgs
+  trajectory_msgs
+)
+ament_export_dependencies(
```
When I enable `ament_export_targets` I get the following during compile. Am I missing something. See it commented out in the new commit.

```
CMake Error in CMakeLists.txt:
  Target "moveit_visual_tools" INTERFACE_INCLUDE_DIRECTORIES property
  contains path:

    "/root/ws_moveit/src/moveit_visual_tools/include"

  which is prefixed in the source directory.


CMake Generate step failed.  Build files cannot be regenerated correctly.
---
Failed   <<< moveit_visual_tools [7.78s, exited with code 1]
```

vatanaksoytezer commented:
I've adressed most issues from the review. Just had trouble with one export (see comment above). @henningkayser feel free to add more comments to my answers.

henningkayser commented on CMakeLists.txt:
```diff
@@ -1,96 +1,116 @@
-  ${catkin_LIBRARIES} ${Boost_LIBRARIES}
+
+target_link_libraries(
+  ${PROJECT_NAME}
```
... and removed from here

henningkayser replied:
nothing is linked here, so this can be removed

henningkayser commented on CMakeLists.txt:
```diff
@@ -54,43 +41,95 @@ add_library(${PROJECT_NAME}
+    include
+)
+ament_target_dependencies(${PROJECT_NAME}
+  Boost
```
nit: use `${THIS_PACKAGE_INCLUDE_DEPENDS}`

henningkayser commented on CMakeLists.txt:
```diff
@@ -54,43 +41,95 @@ add_library(${PROJECT_NAME}
+  ${PROJECT_NAME}
+)
+ament_target_dependencies(${PROJECT_NAME}_demo
+  Boost
```
same

vatanaksoytezer commented:
Fixed the export issue, addressed your last review. This should be (mostly) ready to merge after you test it @henningkayser. Thanks again for such a detailed review, I really appreciate it!

AdamPettinger commented on CMakeLists.txt:
```diff
@@ -54,43 +41,65 @@ add_library(${PROJECT_NAME}
 
 # Install header files
-install(DIRECTORY include/${PROJECT_NAME}/   DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION})
+install(DIRECTORY include DESTINATION include)
```
I am using your branch here on another project, and `#include <moveit_visual_tools/moveit_visual_tools.h>` was failing. On further inspection, `moveit_visual_tools` includes were being installed to `install/moveit_visual_tools/include/include/moveit_visual_tools/` (notice the extra include)

I am not sure exactly why, but adding a `/` during the install command seems to fix this problem, removing the extra `include` in the path, and allowing the other package to find the file

```suggestion
install(DIRECTORY include/ DESTINATION include)
```

AdamPettinger commented:
I've built and tested this locally. The demo runs well, and I was able to use `moveit_visual_tools` in another package successfully (but not used as in depth as the demo of course)

henningkayser commented on README.md:
````diff
@@ -72,7 +76,7 @@ moveit_visual_tools::MoveItVisualToolsPtr visual_tools_;
 In your class' constructor add:
 ```
-visual_tools_.reset(new moveit_visual_tools::MoveItVisualTools("base_frame","/moveit_visual_markers"));
+visual_tools_.reset(new moveit_visual_tools::MoveItVisualTools(node_, "world", "/moveit_visual_tools"));
````
make_shared instead of new?

henningkayser commented on moveit_visual_tools.repos:
```diff
@@ -0,0 +1,9 @@
+    version: foxy-devel
+  graph_msgs:
+    type: git
+    url: https://github.com/vatanaksoytezer/graph_msgs
```
```suggestion
    url: https://github.com/PickNikRobotics/graph_msgs
```

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -147,15 +149,12 @@ bool MoveItVisualTools::processCollisionObjectMsg(const moveit_msgs::CollisionOb
     scene->setObjectColor(msg.id, getColor(color));
   }
   // Trigger an update
-  if (!mannual_trigger_update_)
```
Why is manual mode being removed?

henningkayser commented on src/moveit_visual_tools.cpp:
```diff
@@ -361,25 +358,26 @@ void MoveItVisualTools::loadRobotStatePub(const std::string& robot_state_topic,
   // Wait for topic to be ready
   if (blocking)
-    waitForSubscriber(pub_robot_state_);
+    waitForSubscriber(pub_robot_state_, 5.0 /* seconds */);
```
Throw warning if no subscribers are present

henningkayser commented on src/moveit_visual_tools_demo.cpp:
```diff
@@ -62,26 +68,26 @@ class VisualToolsDemo
+  : node_(node)
   {
-    visual_tools_.reset(new moveit_visual_tools::MoveItVisualTools("world", "/moveit_visual_tools"));
+    visual_tools_.reset(new moveit_visual_tools::MoveItVisualTools(node_, "world", "/moveit_visual_tools"));
```
I mean using `std::make_shared` instead of `reset(new ...)`

henningkayser commented on src/moveit_visual_tools_demo.cpp:
```diff
@@ -62,26 +68,26 @@ class VisualToolsDemo
-    ros::spinOnce();
-    ros::Duration(0.1).sleep();
+    rclcpp::spin_some(node);
+    rclcpp::sleep_for(std::chrono::milliseconds(1000));
```
This changes 100ms to 1000ms. Is that intentional? Also, consider using chrono literals `100ms`

henningkayser replied:
same for following sleeps

vatanaksoytezer replied:
That is intentional, 100ms was way too short to see what was happening in the demo. 1 second seemed better

henningkayser commented on src/moveit_visual_tools_demo.cpp:
```diff
@@ -182,18 +198,17 @@ class VisualToolsDemo
-    std::string file_path = "file://" + ros::package::getPath(THIS_PACKAGE);
+    RCLCPP_INFO_STREAM(LOGGER, "Publishing Collision Mesh");
+    // TODO: Catch exception
+    std::string file_path = "file://" + ament_index_cpp::get_package_share_directory(THIS_PACKAGE);
```
nit, use [get_resource()](https://github.com/ament/ament_index/blob/master/ament_index_cpp/include/ament_index_cpp/get_resource.hpp#L61)?

henningkayser reviewed (commented):
I guess this is the last round of comments ;)

JafarAbdi commented on .clang-format:
```diff
@@ -1,75 +0,0 @@
----
```
Revert .?

JafarAbdi commented on launch/demo_rviz.launch.py:
```diff
@@ -0,0 +1,74 @@
+
+def generate_launch_description():
+
+    robot_description_config = open(
+        os.path.join(
+            get_package_share_directory("moveit_visual_tools"),
+            "resources",
+            "rrbot.urdf",
+        )
+    ).read()
```
Shouldn't we use `load_file` here .?

JafarAbdi commented on src/imarker_end_effector.cpp:
```diff
@@ -141,22 +144,23 @@ bool IMarkerEndEffector::setPoseFromRobotState()
     // Save pose to file if its been long enough
     double save_every_sec = 0.1;
-    if (time_since_last_save_ < ros::Time::now() - ros::Duration(save_every_sec))
+    if (time_since_last_save_ < rclcpp::Clock(RCL_ROS_TIME).now() - rclcpp::Duration(save_every_sec))
```
```suggestion
    if (time_since_last_save_ < rclcpp::Clock(RCL_ROS_TIME).now() - rclcpp::Duration::from_seconds(save_every_sec))
```

JafarAbdi replied:
Also, I think we should have a clock as a member variable rather than constructing it on the fly on each call to this function

JafarAbdi commented on src/imarker_end_effector.cpp:
```diff
@@ -141,22 +144,23 @@ bool IMarkerEndEffector::setPoseFromRobotState()
     {
       imarker_parent_->saveToFile();
-      time_since_last_save_ = ros::Time::now();
+      time_since_last_save_ = rclcpp::Clock(RCL_ROS_TIME).now();
```
See my previous comment

JafarAbdi reviewed (changes requested):
There're few bugs related to `rclcpp::Duration` that need to be fixed before merging this

vatanaksoytezer commented on moveit_visual_tools.repos:
```diff
@@ -0,0 +1,9 @@
+repositories:
+  rviz_visual_tools:
+    type: git
+    url: https://github.com/PickNikRobotics/rviz_visual_tools
+    version: foxy-devel
```
Yes, this needs https://github.com/PickNikRobotics/rviz_visual_tools/pull/185 which has been released. On a related note: I think we should start releasing mvt and rvt regularly within moveit release cycle.

vatanaksoytezer commented on launch/debug_settings.gdb:
```diff
@@ -1,2 +0,0 @@
-set breakpoint pending on
```
I don't think we would need gdb right now, I can add it back though.

vatanaksoytezer commented on src/imarker_robot_state.cpp:
```diff
@@ -98,7 +103,8 @@ bool isIKStateValid(const planning_scene::PlanningScene* planning_scene, bool ve
     planning_scene->isStateColliding(*robot_state, group->getName(), true);
     visual_tools->publishContactPoints(*robot_state, planning_scene);
-    ROS_WARN_STREAM_THROTTLE_NAMED(2.0, "imarker_robot_state", "Collision in IK CC callback");
+    rclcpp::Clock steady_clock(RCL_STEADY_TIME);
```
Making it static seems reasonable, done 👍🏻

vatanaksoytezer commented:
@henningkayser @JafarAbdi is there anything else you would like me to address?
